sensor/kubernetes/eventpipeline/resolver/resolver_bench_test.go

```diff
@@ -0,0 +1,203 @@
+package resolver
+
+import (
+	"fmt"
+	"math/rand"
+	"testing"
+
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/sensor/common/service"
+	"github.com/stackrox/rox/sensor/common/store"
+	mocksStore "github.com/stackrox/rox/sensor/common/store/mocks"
+	"github.com/stackrox/rox/sensor/common/store/resolver"
+	"github.com/stackrox/rox/sensor/kubernetes/eventpipeline/component"
+	mocksComponent "github.com/stackrox/rox/sensor/kubernetes/eventpipeline/component/mocks"
+	"go.uber.org/mock/gomock"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	// lastDeploymentID is used to mark when we should stop waiting for the deployments to be processed.
+	lastDeploymentID = "last-deployment-id"
+	// queueSize is the innerQueue size.
+	queueSize = 100
+)
+
+var (
+	res                 component.Resolver
+	mockCtrl            *gomock.Controller
+	mockOutput          *mocksComponent.MockOutputQueue
+	mockDeploymentStore *mocksStore.MockDeploymentStore
+	mockServiceStore    *mocksStore.MockServiceStore
+	mockRBACStore       *mocksStore.MockRBACStore
+	mockEndpointManager *mocksStore.MockEndpointManager
+
+	cases = []struct {
+		numEvents      int
+		numDeployments int
+	}{
+		{
+			numEvents:      100,
+			numDeployments: 1,
+		},
+		{
+			numEvents:      100,
+			numDeployments: 10,
+		},
+		{
+			numEvents:      1000,
+			numDeployments: 100,
+		},
+		{
+			numEvents:      1000,
+			numDeployments: 1000,
+		},
+	}
+)
+
+func BenchmarkProcessDeploymentReferences(b *testing.B) {
+	for _, bc := range cases {
+		b.Run(fmt.Sprintf("Benchmark with %d events and %d deployments per event", bc.numEvents, bc.numDeployments), func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				b.StopTimer()
+				doneSignal := concurrency.NewSignal()
+				setupMocks(b, &doneSignal)
+				events := createEvents(false, bc.numEvents, bc.numDeployments)
+				setupResolver(b)
+				b.StartTimer()
+				for _, event := range events {
+					res.Send(event)
+				}
+				doneSignal.Wait()
+				b.StopTimer()
+			}
+		})
+	}
+}
+
+func BenchmarkProcessRandomDeploymentReferences(b *testing.B) {
+	for _, bc := range cases {
+		b.Run(fmt.Sprintf("Benchmark with %d events and %d random deployments per event", bc.numEvents, bc.numDeployments), func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				b.StopTimer()
+				doneSignal := concurrency.NewSignal()
+				setupMocks(b, &doneSignal)
+				events := createEvents(true, bc.numEvents, bc.numDeployments)
+				setupResolver(b)
+				b.StartTimer()
+				for _, event := range events {
+					res.Send(event)
+				}
+				doneSignal.Wait()
+				b.StopTimer()
+			}
+		})
+	}
+}
+
+func setupResolver(b *testing.B) {
+	res = New(mockOutput, &fakeProvider{
+		deploymentStore: mockDeploymentStore,
+		serviceStore:    mockServiceStore,
+		rbacStore:       mockRBACStore,
+		endpointManager: mockEndpointManager,
+	}, queueSize)
+	err := res.Start()
```

Comment:
Start is called many times (i.e. everytime `setupResolver` is called), shouldn't you call `res.Stop()` at the end of every test case? Otherwise there will be leaking goroutines that might affect other test cases.

Reply:
😱 I thought I added that! Yes, the resolver should be stopped otherwise we have many resolvers running at the same time. It shouldn't affect to much the benchmark because we only send messages to the one on scope but still...